test/unit/spec/PrimaryAuth_spec.js

```diff
@@ -885,6 +888,72 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
       });
     });
 
+    Expect.describe('Typing biometrics', function () {
+      itp('does not contain typing pattern header in primary auth request if feature is disabled', function () {
+        return setup({ features: { trackTypingPattern: false }})
+        .then(function (test) {
+          $.ajax.calls.reset();
+          test.form.setUsername('testuser');
```

Comment:
Can you also add a test for what X-Typing-Pattern header contains when username is null and when username is just a space char? Or will the sign-in widget catch these cases before getTypingPattern() gets called?

Reply:
when username is null form cannot be submitted so as you said sign-in widget catch these cases before getTypingPattern() gets called. When username is just a space it would be treated as another character and typing pattern would be computed.  I am not seeing a need to testing this unless I am missing something.

Reply:
Just curious if X-Typing-Pattern will have an empty value or have the corresponding pattern for the space character. What would be better I think is if the sign-in widget trimmed the trailing space, caught the empty username case, and not need to make subsequent calls to getTypingPattern(), etc.

Reply:
According to the TypingDNA api and what I've seen, the X-Typing-Pattern header can only produce either one of 'null' or the typing pattern. If there is no typing information that can be caught (like when username is empty or autofilled), TypingDNA produces 'null'.

Reply:
@ruthadrineda-okta  agreed! I think limiting the submission of a space should be considered in general even in courage. (Will bring it up in UI office hours and create a JIRA if required). But outside the scope of this JIRA. For now space isa  valid character and honestly getTypingPattern is not expensive that we need to worry about it for typing patterns. 
https://oktawiki.atlassian.net/wiki/spaces/eng/pages/236645615/UICore+Office+Hours+Meeting+Notes